src/lib/proxy-server.ts

```diff
@@ -0,0 +1,189 @@
+			const host = req.headers.host?.split( ':' )[ 0 ]; // Remove port if present
+
+			// Log incoming request information for debugging
+			console.log( `Received request with host header: ${ host }` );
+
+			// Look up the port directly from user data
+			if ( ! host ) {
+				console.log( 'No host header found' );
+				res.writeHead( 404 );
+				res.end( 'No host header found' );
+				return;
+			}
+
+			const port = await getPortForDomain( host );
+			if ( ! port ) {
+				console.log( `Domain not found: ${ host }` );
+				res.writeHead( 404 );
+				res.end( `Domain not found: ${ host }` );
+				return;
+			}
+
+			console.log( `Proxying request for ${ host } to port ${ port }` );
+
+			// Forward the request with the original host preserved
+			proxy.web( req, res, {
+				target: `http://localhost:${ port }`,
+				xfwd: true, // Pass along x-forwarded headers
+			} );
+		} );
+
+		await new Promise< void >( ( resolve, reject ) => {
+			proxyServer!
+				.listen( port, () => {
+					console.log( `Proxy server started on port ${ port }` );
+					isProxyRunning = true;
+					resolve();
+				} )
+				.on( 'error', ( err ) => {
+					console.error( `Error starting proxy server on port ${ port }:`, err );
+					reject( err );
+				} );
+		} );
+
+		return true;
+	} catch ( error ) {
+		console.error( `Failed to start proxy server directly:`, error );
+		return false;
+	}
+}
+
+/**
+ * Attempts to start the proxy server with elevated privileges
+ */
+async function startProxyWithElevation(): Promise< boolean > {
+	const currentPlatform = platform();
+
+	try {
+		console.log( 'Attempting to start proxy server with elevated privileges' );
+
+		if ( currentPlatform === 'win32' ) {
+			// For Windows: First start the actual proxy on a non-privileged port (8880)
+			const proxyStarted = await startDomainProxy( 8880 );
+			if ( ! proxyStarted ) {
+				console.error( 'Failed to start proxy server on port 8880' );
+				return false;
+			}
+
+			// Then use netsh to forward from privileged port 80 to our proxy on 8880
+			const command =
+				'netsh interface portproxy add v4tov4 listenport=80 listenaddress=127.0.0.1 connectport=8880 connectaddress=127.0.0.1';
+
+			try {
+				// @ts-expect-error promisify doesn't seem typed properly.
+				await sudoExec( command, { name: 'WordPress Studio Proxy' } );
+				console.log( 'Successfully set up port forwarding from port 80 to 8880' );
+				return true;
+			} catch ( error ) {
+				console.error( 'Failed to set up port forwarding:', error );
+				// Even though port forwarding failed, we still have a working proxy on 8880
+				return proxyStarted;
+			}
+		} else {
+			// For Mac/Linux: Use sudo to run a command that starts a proxy directly on port 80
+			// This is a more complex solution that would require a proper script
+			// For now, we'll just return false and let the app fall back to non-privileged ports
+			console.log( 'Elevated privileges for Mac/Linux not fully implemented yet' );
```

Comment:
In MacOS I didn't really need elevated privileges to run the proxy in my tests. But worth testing more.